geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/process/JoinProcess.scala

```diff
@@ -0,0 +1,250 @@
+      val attributes = Seq(joinAttribute) ++ primaryAttributes ++ secondaryAttributes
+      getCombinedSft(primary.getSchema, secondary.getSchema, attributes, joinAttribute)
+    }
+
+    // check for too many features coming back - limit is somewhat arbitrary, but this
+    // class is mainly intended for a single feature lookup
+    val primaryFeatures = SelfClosingIterator(primary).toList
+    require(primaryFeatures.length < 129,
+      s"Too many features returned from primary query - got ${primaryFeatures.length}, max 128")
+
+    val joinProperty = ff.property(joinAttribute)
+    val joinFilters = {
+      val values = primaryFeatures.map(_.getAttribute(joinAttribute)).distinct
+      values.map(p => ff.equals(joinProperty, ff.literal(p)))
+    }
+
+    val result = if (joinFilters.isEmpty) {
+      new DefaultFeatureCollection(null, returnSft)
+    } else {
+      val or = ff.or(joinFilters.toList)
+      val filter = if (joinFilter != null && joinFilter != Filter.INCLUDE) { ff.and(or, joinFilter) } else { or }
+      val visitor = new QueryVisitor(secondary, filter)
+      secondary.accepts(visitor, new NullProgressListener)
+      val results = visitor.getResult.asInstanceOf[QueryResult].results
+
+      // mappings from the secondary feature result to the return schema
+      // (return sft index, result sft index, is from primary result (or secondary result))
+      val attributeMappings: Seq[(Int, Int, Boolean)] = returnSft.getAttributeDescriptors.map { d =>
+        val toAttribute = d.getLocalName
+        val dot = toAttribute.indexOf('.')
+        if (dot == -1) {
+          val fromPrimary = secondary.getSchema.getAttributeDescriptors.exists(_.getLocalName == toAttribute)
+          val from = if (fromPrimary) {
+            primary.getSchema.indexOf(toAttribute)
+          } else {
+            secondary.getSchema.indexOf(toAttribute)
+          }
+          (returnSft.indexOf(toAttribute), from, fromPrimary)
+        } else {
+          val fromPrimary = primary.getSchema.getTypeName == toAttribute.substring(0, dot)
+          val from = if (fromPrimary) {
+            primary.getSchema.indexOf(toAttribute.substring(dot + 1))
+          } else {
+            secondary.getSchema.indexOf(toAttribute.substring(dot + 1))
+          }
+          (returnSft.indexOf(toAttribute), from, fromPrimary)
+        }
+      }
+
+      new DecoratingSimpleFeatureCollection(results) {
+        override def getSchema: SimpleFeatureType = returnSft
+        override def features(): SimpleFeatureIterator = new SimpleFeatureIterator {
+          val delegate = results.features
+          override def next(): SimpleFeature = {
+            val secondarySf = delegate.next()
+            val toJoin = secondarySf.getAttribute(joinAttribute)
+            val primarySf = primaryFeatures.find(_.getAttribute(joinAttribute) == toJoin).getOrElse {
+              throw new RuntimeException("No feature joined from attribute query")
```

Comment:
shouldn't be possible for there to be any misses. The join filter is created based on the attribute values from the primary features, so each join feature should match at least one primary feature.